securedrop/debian/config/usr/bin/securedrop-noble-migration-check.py

```diff
@@ -0,0 +1,137 @@
+            if line.startswith("VERSION_CODENAME="):
+                return line.split("=")[1].strip().strip('"')
+    raise ValueError("Could not determine Ubuntu codename")
+
+
+def check_ssh_group() -> bool:
+    """Check the UNIX "ssh" group is empty"""
+    try:
+        group = grp.getgrnam("ssh")
+    except KeyError:
+        print("ssh: group does not exist")
+        return True
+
+    if len(group.gr_mem) > 0:
+        print(f"ssh: group is not empty: {group.gr_mem}")
+        return False
+
+    print("ssh: group is empty")
+    return True
+
+
+def check_ufw_removed() -> bool:
+    """Check that ufw is removed"""
+    if Path("/usr/sbin/ufw").exists():
+        print("ufw: ufw is still installed")
+        return False
+    else:
+        print("ufw: ufw was removed")
+        return True
+
+
+def check_free_space() -> bool:
+    """Check the root partition has more than 10G free"""
+    root = shutil.disk_usage("/")
+    if root.free < (10 * 1024 * 1024 * 1024):
+        print("free space: not enough free space")
+        return False
+    else:
+        print("free space: enough free space")
+        return True
+
+
+def check_apt() -> bool:
+    """Verify only expected sources are configured"""
+    try:
+        output = subprocess.check_output(["apt-get", "indextargets"], text=True)
+    except subprocess.CalledProcessError:
+        print("apt: error invoking apt-get indextargets")
+        return False
+
+    for line in output.splitlines():
+        if line.startswith("URI:"):
+            parsed = urlparse(line.split(":", 1)[1].strip())
+            if parsed.hostname not in [
+                "archive.ubuntu.com",
+                "security.ubuntu.com",
+                "apt.freedom.press",
+                "apt-test.freedom.press",
+            ]:
+                print(f"apt: unexpected source: {parsed.hostname}")
+                return False
+
+    print("apt: all sources are expected")
+    return True
+
+
+def check_systemd() -> bool:
+    """ "No systemd units are failed"""
+    try:
+        subprocess.check_output(["systemctl", "is-failed"])
+        print("systemd: some systemd units are failed")
+        return False
+    except subprocess.CalledProcessError:
+        print("systemd: all systemd units are running")
+        return True
+
+
+def main() -> None:
+    if os.geteuid() != 0:
+        print("You need to run this as root")
+        sys.exit(1)
+
+    state_path = Path("/etc/securedrop-noble-migration.json")
+    codename = os_codename()
+
+    if codename != "focal":
+        print(f"Unsupported Ubuntu version: {codename}")
+        # nothing to do, write an empty JSON blob
+        state_path.write_text(json.dumps({}))
+        return
+
+    state = {}
+    state["ssh"] = check_ssh_group()
+    state["ufw"] = check_ufw_removed()
+    state["free_space"] = check_free_space()
+    state["apt"] = check_apt()
+    state["systemd"] = check_systemd()
+
+    state_path.write_text(json.dumps(state))
```

Comment:
We need some crash resistance here; if something fails before we dump to disk, we try our hardest to not leave the old state behind.